File: app/src/main/java/org/eu/droid_ng/wellbeing/FocusModeActivity.java
package org.eu.droid_ng.wellbeing;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import android.animation.LayoutTransition;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.eu.droid_ng.wellbeing.lib.GlobalWellbeingState;
import org.eu.droid_ng.wellbeing.lib.WellbeingStateClient;

import java.util.concurrent.atomic.AtomicBoolean;

public class FocusModeActivity extends AppCompatActivity {
	private WellbeingStateClient client;
	private Runnable next;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		ActionBar actionBar = getSupportActionBar();
		if (actionBar != null) {
			actionBar.setDisplayHomeAsUpEnabled(true);
		}
		client = new WellbeingStateClient(this);
		setContentView(R.layout.activity_focusmode);
		AtomicBoolean focusMode = new AtomicBoolean(false);
		AtomicBoolean focusModeBreak = new AtomicBoolean(false);
		LayoutTransition layoutTransition = ((LinearLayout) findViewById(R.id.focusModeRoot)).getLayoutTransition();
		layoutTransition.enableTransitionType(LayoutTransition.CHANGING);
		next = () -> {
			Button toggle = findViewById(R.id.focusModeToggle);
			TextView takeBreak = findViewById(R.id.focusModeBreak);
			if (!focusMode.get()) {
				toggle.setText(R.string.enable);
				toggle.setOnClickListener(v -> client.doBindService(state -> {
					state.stateChangeCallbacks.add(() -> {
						focusMode.set(state.type == GlobalWellbeingState.SERVICE_TYPE.TYPE_FOCUS_MODE);
						focusModeBreak.set(state.focusModeBreak);
						next.run();
					});
					state.enableFocusMode();
				}, false, true, false));
				takeBreak.setVisibility(View.GONE);
				takeBreak.setOnClickListener(null);
			} else {
				toggle.setText(R.string.disable);
				toggle.setOnClickListener(v -> client.doBindService(GlobalWellbeingState::disableFocusMode, false, true, true));
				takeBreak.setVisibility(View.VISIBLE);
				if (focusModeBreak.get()) {
					takeBreak.setOnClickListener(v -> client.doBindService(GlobalWellbeingState::endBreak));
					takeBreak.setText(R.string.focus_mode_break_end);
				} else {
					takeBreak.setOnClickListener(v -> client.doBindService(state -> state.takeBreakDialog(FocusModeActivity.this, false, null)));
					takeBreak.setText(R.string.focus_mode_break);
				}
			}
		};
		if (!client.isServiceRunning()) {
			focusMode.set(false);
			focusModeBreak.set(false);
			next.run();
		} else {
			client.doBindService(state -> {
				Runnable r = () -> {
					focusMode.set(state.type == GlobalWellbeingState.SERVICE_TYPE.TYPE_FOCUS_MODE);
					focusModeBreak.set(state.focusModeBreak);
					next.run();
				};
				state.stateChangeCallbacks.add(r);
				r.run();
			});
		}

		TextView schedule = findViewById(R.id.schedule);
		schedule.setOnClickListener(v -> {
			startActivity(new Intent(this, ScheduleActivity.class).putExtra("type", "focus_mode"));
		});

		RecyclerView r = findViewById(R.id.focusModePkgs);
		r.setAdapter(
				new PackageRecyclerViewAdapter(this,
						getPackageManager().getInstalledApplications(PackageManager.GET_META_DATA),
						"focus_mode"));
	}

	@Override
	public boolean onSupportNavigateUp() {
		finish();
		return true;
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		client.doUnbindService();
	}
}

File: app/src/main/java/org/eu/droid_ng/wellbeing/ui/FocusModeActivity.kt
package org.eu.droid_ng.wellbeing.ui

import android.animation.LayoutTransition
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.LinearLayoutCompat
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.appbar.MaterialToolbar
import org.eu.droid_ng.wellbeing.R
import org.eu.droid_ng.wellbeing.lib.WellbeingService
import org.eu.droid_ng.wellbeing.prefs.PackageRecyclerViewAdapter
import org.eu.droid_ng.wellbeing.widget.MainSwitchBar
import java.util.function.Consumer

class FocusModeActivity : AppCompatActivity() {

    private val service: WellbeingService by lazy {
        WellbeingService.get()
    }

    private val stateCallback = Consumer<WellbeingService> { _: WellbeingService ->
        updateUi()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_focus_mode)
        // Set support action bar
        val topAppBar = findViewById<MaterialToolbar>(R.id.topbar)
        setSupportActionBar(topAppBar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        // Enable layout transition
        val layoutTransition =
            (findViewById<View>(R.id.focusModeRoot) as LinearLayoutCompat).layoutTransition
        layoutTransition.enableTransitionType(LayoutTransition.CHANGING)
        // Open schedule screen
//        findViewById<View>(R.id.schedule)?.setOnClickListener {
//            startActivity(
//                Intent(
//                    this,
//                    ScheduleActivity::class.java
//                ).putExtra("type", "focus_mode").putExtra("name", getString(R.string.focus_mode))
//            )
//        }
        // Add state call back to the service
        service.addStateCallback(stateCallback)
        val r = findViewById<RecyclerView>(R.id.focusModePkgs)
        r.adapter = PackageRecyclerViewAdapter(
            this,
            service.getInstalledApplications(PackageManager.GET_META_DATA),
            "focus_mode"
        ) { packageName: String? ->
            service.onFocusModePreferenceChanged(
                packageName!!
            )
        }
        // Handle on click for main switch
        findViewById<MainSwitchBar>(R.id.mainSwitchBar)?.setOnClickListener {
            val state = service.getState()
            if (state.isFocusModeEnabled()) {
                service.disableFocusMode()
            } else {
                service.enableFocusMode()
            }
        }

        updateUi()
    }

    override fun onDestroy() {
        super.onDestroy()
        service.removeStateCallback(stateCallback)
    }

    private fun updateUi() {
        val state = service.getState()
        updateMainSwitch(state.isFocusModeEnabled())
//        val takeBreak = findViewById<View>(R.id.takeBreak)
//        (findViewById<View>(R.id.title) as AppCompatTextView).setText(if (state.isOnFocusModeBreakGlobal()) R.string.focus_mode_break_end else R.string.focus_mode_break)
//        takeBreak.setOnClickListener { v: View? ->
//            if (state.isOnFocusModeBreakGlobal()) {
//                service.endFocusModeBreak()
//            } else {
//                service.takeFocusModeBreakWithDialog(this@FocusModeActivity, false, null)
//            }
//        }
//        takeBreak.visibility = if (state.isFocusModeEnabled()) View.VISIBLE else View.GONE
    }

    private fun updateMainSwitch(checked: Boolean) {
        findViewById<MainSwitchBar>(R.id.mainSwitchBar)?.isChecked = checked
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }
}
